### PR TITLE
Feat: Add definitions for embedded languages

### PR DESCRIPTION
**Pull request by idillon-sfl**
Note some definitions do not work properly, such as the ones for the `utils` and `contains` parts of `bb.utils.contains`. These are the sames the same for which hover is deffective.

I tried something in order to handle `d` and `e`. I am not sure this is the desired behavior. Please tell me what you think of it.

### CODE REVIEW

File: client/src/language/middlewareDefinition.ts [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+      if (newTargetSelectionRange !== undefined) {
+        definition.targetSelectionRange = newTargetRange
+      }
+    }
+  }
+}
+
+// Redirect a definition to an other definition
+// For example, `d` of `d.getVar('')` is redirected to the definition of `data_smart.DataSmart()`
+const redirectDefinition = async <DefinitionType extends Location | LocationLink>(
```
Comment (deribaucourt) [inline client/src/language/middlewareDefinition.ts:110]: Go to definition works properly in the temporary. Except that yes, as can be expected, d goes to the d assignment and not the definition of the class. So it makes sense to instead point to the definition of DataSmart(). Also with your code, d.getVar points correctly to the getVar function. Perfect!

File: client/src/language/middlewareDefinition.ts [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+): Promise<DefinitionType[]> => {
+  const result: DefinitionType[] = []
+  await Promise.all(definitions.map(async (definition) => {
+    if (!checkIsDefinitionUriEqual(definition, Uri.parse(embeddedLanguageDocInfos.uri))) {
+      result.push(definition) // only definitions located on the embedded language documents need ajustments
+      return
+    }
+    if (embeddedLanguageDocInfos.language === 'python') {
+      const dDefinition = await getDefinitionOfD(definition)
+      if (dDefinition !== undefined) {
```
Comment (deribaucourt) [inline client/src/language/middlewareDefinition.ts:65]: Could look a bit more tidy if we use a dictionary here? keys would be the symbol of which we are swapping the definition, and value could be the location of the definition? Would look better if we add more of them. (oe?)

File: client/src/language/utils.ts [hunk 3]
```diff
@@ -3,7 +3,7 @@
  * Licensed under the MIT License. See License.txt in the project root for license information.
  * ------------------------------------------------------------------------------------------ */
 
-import { Position, Range, type TextDocument } from 'vscode'
+import { Location, type LocationLink, Position, Range, type Uri, type TextDocument } from 'vscode'
```
Comment (deribaucourt) [inline client/src/language/utils.ts:6]: Could you add a debug log with the location of temporary files? Would make it easier to browse them when debugging embedded languages features.

File: client/src/language/middlewareDefinition.ts [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    position
+  )
+  const vdocUri = Uri.parse(embeddedLanguageDocInfos.uri)
+  const tempResult = await commands.executeCommand<Location[] | LocationLink[]>(
+    'vscode.executeDefinitionProvider',
+    vdocUri,
+    adjustedPosition
+  )
+
+  if (checkIsArrayLocation(tempResult)) {
```
Comment (deribaucourt) [inline client/src/language/middlewareDefinition.ts:40]: Are both branches executing the same code?

**Review by deribaucourt — CHANGES_REQUESTED** [review]

File: client/src/language/middlewareDefinition.ts (hunk 2 above)
Comment (idillon-sfl) [reply client/src/language/middlewareDefinition.ts]: Makes sense. I did something in the kind.

File: client/src/language/utils.ts (hunk 3 above)
Comment (idillon-sfl) [reply client/src/language/utils.ts:6]: done

File: client/src/language/middlewareDefinition.ts (hunk 4 above)
Comment (idillon-sfl) [reply client/src/language/middlewareDefinition.ts:46]: Yes. They also produce the same compiled javascript code. This check is only to please TypeScript. I'd rather have this pointless check than losing the type assurance provided by TypeScript. That being said, if you look deeper, you'll see all the functions I added in utils.ts begin with `if (definition instanceof Location)`. It shows there could have been two separate branches, but I thought it was easier reason on a single algorithm that does not mind whether the definition is "Location" or "LocationLink".

However, there are still two type assertions in [convertToSameDefinitionType](https://github.com/yoctoproject/vscode-bitbake/pull/38/files#diff-a5ff141411e32f5d167aad66a4678463d637f38307fe7dc22d0ec1cf1ee5109cR128). This is the best I could do.

**Review by deribaucourt — CHANGES_REQUESTED** [review]
The newest patchset does not seem to work on my setup.
Any time I hover onto embedded functions, nothing happens except the problems tab flashes some errors very quickly. Looks like the embedded document gets properly handled but immediately disabled.

**Review by deribaucourt — DISMISSED** [review]
Doesn't work inside functions, but that's expected with dynamic Python typing.

**Comment by idillon-sfl** [comment]
definition.test.ts and command-wrapper.test.ts were both opening meta-fixtures/definition.bb. For reasons that are not well understood, it was making the tests fail. Now they both open a different file.

**Review by deribaucourt — APPROVED** [review]

**Pull request merged**